# Capstone Project - The Battle of Neighborhoods

## Introduction

Over the last five years, crime rates in London have increased by over 23%. A total of 15,590 knife crimes were reported across London from 2019-2020. The concentration of crimes across different London boroughs should be considered by commercial and residential real estate agents, as well as anybody looking to buy or rent a home in the city. Hence, this report is of interest to anyone that fulfils such criteria. 

The business problem we are faced with is: which parts of London are safe and attractive for residential buyers/renters across various demographic groups. To solve this problem, we will assess the crime rates in each of the London boroughs, and cluster neighborhoods to assess the venues on offer, such as pubs, cafes and parks.

## Data

The data that will be used to assess the best residential areas in London is: (1) the Metropolitan Police Service (MPS) Borough Level Crime from August 2018 to July 2020 (https://data.london.gov.uk), and (2) Foursquare venue data.

The headings from the MPS Borough Level Crime data are: MajorText, MinorText, LookUp_Borough and 24 separate columns for the 24 months of crime data. The major and minor text describe the accused crime, for example “violence against the person” and “violence with injury”. In order to analyse this data, we must sum the 24 columns to find the total crime levels in each borough. We must then group the data by Borough, which we will refer to as the ‘Address’.

## Methodology

The methodology section will consist of five sections covering the exploratory data analysis, statistical testing and machine learning involved in the project.

I. Downloading and exploring the dataset

II.	Exploring neighborhoods in London

III. Analysing each neighbourhood

IV. Clustering neighborhoods

V. Examining clusters

In [1]:
import numpy as np
import pandas as pd
import datetime as dt # Datetime
import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
print('Libraries imported.')

Libraries imported.


### Downloading and exploring the dataset

Loading the data and transforming it into a *pandas* dataframe

In [2]:
# Read the data for examination (Source: https://data.london.gov.uk/dataset/recorded_crime_summary)
df_mps = pd.read_csv("https://data.london.gov.uk/download/recorded_crime_summary/d2e9ccfc-a054-41e3-89fb-53c2bc3ed87a/MPS%20Borough%20Level%20Crime%20%28most%20recent%2024%20months%29.csv")

In [3]:
df_mps.head(5)

MajorText                          MinorText  \
0  Arson and Criminal Damage                              Arson   
1  Arson and Criminal Damage                    Criminal Damage   
2                   Burglary  Burglary - Business and Community   
3                   Burglary             Burglary - Residential   
4              Drug Offences                   Drug Trafficking   

     LookUp_BoroughName  201808  201809  201810  201811  201812  201901  \
0  Barking and Dagenham       5       3       8       5       1       5   
1  Barking and Dagenham     101     107     132     105      88      97   
2  Barking and Dagenham      18      33      32      39      33      45   
3  Barking and Dagenham      84      99      94     106     164     114   
4  Barking and Dagenham       7      10       9       7       4       5   

   201902  ...  201910  201911  201912  202001  202002  202003  202004  \
0       2  ...       9       8       6       4       5       6       2   
1     127  ...     109      97     121      97     103     107      80   
2      24  ...      30      30      25      31      17      28      29   
3     107  ...      97     114     130     116     123      97      57   
4       2  ...       8      12       3      14       5       6      12   

   202005  202006  202007  
0       2       4       3  
1      86     121     121  
2      16      16      28  
3      41      63      72  
4      13      11      20  

[5 rows x 27 columns]

In [4]:
df_mps.shape

(1568, 27)

Renaming the columns

In [5]:
# Assign meaningful column names
df_mps.columns = ['Major_Crime', 'Minor_Crime', 'Address', 'Jul_18', 'Aug_18', 'Sep_18', 'Oct_18', 'Nov_18',\
                  'Dec_18', 'Jan_19', 'Feb_19', 'Mar_19', 'Apr_19', 'May_19', 'Jun_19', 'Jul_19', 'Aug_19',\
                  'Sep_19', 'Oct_19', 'Nov_19', 'Dec_19', 'Jan_20', 'Feb_20', 'Mar_20', 'Apr_20', 'May_20', 'Jun_20']

In [6]:
df_mps

Major_Crime  \
0                Arson and Criminal Damage   
1                Arson and Criminal Damage   
2                                 Burglary   
3                                 Burglary   
4                            Drug Offences   
5                            Drug Offences   
6     Miscellaneous Crimes Against Society   
7     Miscellaneous Crimes Against Society   
8     Miscellaneous Crimes Against Society   
9     Miscellaneous Crimes Against Society   
10    Miscellaneous Crimes Against Society   
11    Miscellaneous Crimes Against Society   
12    Miscellaneous Crimes Against Society   
13    Miscellaneous Crimes Against Society   
14    Miscellaneous Crimes Against Society   
15    Miscellaneous Crimes Against Society   
16    Miscellaneous Crimes Against Society   
17    Miscellaneous Crimes Against Society   
18    Miscellaneous Crimes Against Society   
19    Miscellaneous Crimes Against Society   
20    Miscellaneous Crimes Against Society   
21    Miscellaneous Crimes Against Society   
22    Miscellaneous Crimes Against Society   
23    Miscellaneous Crimes Against Society   
24                   Possession of Weapons   
25                   Possession of Weapons   
26                   Possession of Weapons   
27                   Possession of Weapons   
28                   Possession of Weapons   
29                   Public Order Offences   
...                                    ...   
1538  Miscellaneous Crimes Against Society   
1539  Miscellaneous Crimes Against Society   
1540  Miscellaneous Crimes Against Society   
1541  Miscellaneous Crimes Against Society   
1542  Miscellaneous Crimes Against Society   
1543  Miscellaneous Crimes Against Society   
1544                 Possession of Weapons   
1545                 Possession of Weapons   
1546                 Possession of Weapons   
1547                 Possession of Weapons   
1548                 Possession of Weapons   
1549                 Public Order Offences   
1550                 Public Order Offences   
1551                 Public Order Offences   
1552                 Public Order Offences   
1553                               Robbery   
1554                               Robbery   
1555                       Sexual Offences   
1556                       Sexual Offences   
1557                                 Theft   
1558                                 Theft   
1559                                 Theft   
1560                                 Theft   
1561                      Vehicle Offences   
1562                      Vehicle Offences   
1563                      Vehicle Offences   
1564                      Vehicle Offences   
1565           Violence Against the Person   
1566           Violence Against the Person   
1567           Violence Against the Person   

                                            Minor_Crime               Address  \
0                                                 Arson  Barking and Dagenham   
1                                       Criminal Damage  Barking and Dagenham   
2                     Burglary - Business and Community  Barking and Dagenham   
3                                Burglary - Residential  Barking and Dagenham   
4                                      Drug Trafficking  Barking and Dagenham   
5                                   Possession of Drugs  Barking and Dagenham   
6                                         Bail Offences  Barking and Dagenham   
7                                                Bigamy  Barking and Dagenham   
8                                     Dangerous Driving  Barking and Dagenham   
9     Disclosure, Obstruction, False or Misleading S...  Barking and Dagenham   
10                         Exploitation of Prostitution  Barking and Dagenham   
11      Fraud or Forgery Associated with Driver Records  Barking and Dagenham   
12                          Going Equipped for Stealing  Barking and Dagenham   
13                                Handling S

Finding the sum of the 24 months of crime data

In [7]:
# Create total of number columns
df_mps['Total_Crime'] = df_mps.sum(numeric_only=True, axis=1)

In [8]:
df_mps

Major_Crime  \
0                Arson and Criminal Damage   
1                Arson and Criminal Damage   
2                                 Burglary   
3                                 Burglary   
4                            Drug Offences   
5                            Drug Offences   
6     Miscellaneous Crimes Against Society   
7     Miscellaneous Crimes Against Society   
8     Miscellaneous Crimes Against Society   
9     Miscellaneous Crimes Against Society   
10    Miscellaneous Crimes Against Society   
11    Miscellaneous Crimes Against Society   
12    Miscellaneous Crimes Against Society   
13    Miscellaneous Crimes Against Society   
14    Miscellaneous Crimes Against Society   
15    Miscellaneous Crimes Against Society   
16    Miscellaneous Crimes Against Society   
17    Miscellaneous Crimes Against Society   
18    Miscellaneous Crimes Against Society   
19    Miscellaneous Crimes Against Society   
20    Miscellaneous Crimes Against Society   
21    Miscellaneous Crimes Against Society   
22    Miscellaneous Crimes Against Society   
23    Miscellaneous Crimes Against Society   
24                   Possession of Weapons   
25                   Possession of Weapons   
26                   Possession of Weapons   
27                   Possession of Weapons   
28                   Possession of Weapons   
29                   Public Order Offences   
...                                    ...   
1538  Miscellaneous Crimes Against Society   
1539  Miscellaneous Crimes Against Society   
1540  Miscellaneous Crimes Against Society   
1541  Miscellaneous Crimes Against Society   
1542  Miscellaneous Crimes Against Society   
1543  Miscellaneous Crimes Against Society   
1544                 Possession of Weapons   
1545                 Possession of Weapons   
1546                 Possession of Weapons   
1547                 Possession of Weapons   
1548                 Possession of Weapons   
1549                 Public Order Offences   
1550                 Public Order Offences   
1551                 Public Order Offences   
1552                 Public Order Offences   
1553                               Robbery   
1554                               Robbery   
1555                       Sexual Offences   
1556                       Sexual Offences   
1557                                 Theft   
1558                                 Theft   
1559                                 Theft   
1560                                 Theft   
1561                      Vehicle Offences   
1562                      Vehicle Offences   
1563                      Vehicle Offences   
1564                      Vehicle Offences   
1565           Violence Against the Person   
1566           Violence Against the Person   
1567           Violence Against the Person   

                                            Minor_Crime               Address  \
0                                                 Arson  Barking and Dagenham   
1                                       Criminal Damage  Barking and Dagenham   
2                     Burglary - Business and Community  Barking and Dagenham   
3                                Burglary - Residential  Barking and Dagenham   
4                                      Drug Trafficking  Barking and Dagenham   
5                                   Possession of Drugs  Barking and Dagenham   
6                                         Bail Offences  Barking and Dagenham   
7                                                Bigamy  Barking and Dagenham   
8                                     Dangerous Driving  Barking and Dagenham   
9     Disclosure, Obstruction, False or Misleading S...  Barking and Dagenham   
10                         Exploitation of Prostitution  Barking and Dagenham   
11      Fraud or Forgery Associated with Driver Records  Barking and Dagenham   
12                          Going Equipped for Stealing  Barking and Dagenham   
13                                Handling S

Dropping the non-essential columns and creating a new dataframe

In [9]:
df_mps2 = df_mps.drop(columns=['Major_Crime', 'Minor_Crime', 'Jul_18', 'Aug_18', 'Sep_18', 'Oct_18', 'Nov_18',\
                  'Dec_18', 'Jan_19', 'Feb_19', 'Mar_19', 'Apr_19', 'May_19', 'Jun_19', 'Jul_19', 'Aug_19',\
                  'Sep_19', 'Oct_19', 'Nov_19', 'Dec_19', 'Jan_20', 'Feb_20', 'Mar_20', 'Apr_20', 'May_20', 'Jun_20'])

In [10]:
df_mps2

Address  Total_Crime
0     Barking and Dagenham          116
1     Barking and Dagenham         2681
2     Barking and Dagenham          681
3     Barking and Dagenham         2301
4     Barking and Dagenham          199
5     Barking and Dagenham         2362
6     Barking and Dagenham            1
7     Barking and Dagenham            1
8     Barking and Dagenham           29
9     Barking and Dagenham            1
10    Barking and Dagenham            1
11    Barking and Dagenham            8
12    Barking and Dagenham           15
13    Barking and Dagenham           49
14    Barking and Dagenham           19
15    Barking and Dagenham          209
16    Barking and Dagenham           20
17    Barking and Dagenham           35
18    Barking and Dagenham            1
19    Barking and Dagenham           34
20    Barking and Dagenham           14
21    Barking and Dagenham           21
22    Barking and Dagenham          122
23    Barking and Dagenham            2
24    Barking and Dagenham            2
25    Barking and Dagenham          146
26    Barking and Dagenham           19
27    Barking and Dagenham           39
28    Barking and Dagenham          160
29    Barking and Dagenham          326
...                    ...          ...
1538           Westminster          134
1539           Westminster           31
1540           Westminster           49
1541           Westminster           37
1542           Westminster           42
1543           Westminster           94
1544           Westminster            3
1545           Westminster          371
1546           Westminster           23
1547           Westminster          148
1548           Westminster          327
1549           Westminster          692
1550           Westminster         4211
1551           Westminster         1619
1552           Westminster           33
1553           Westminster          352
1554           Westminster         6952
1555           Westminster         1431
1556           Westminster          759
1557           Westminster         3017
1558           Westminster        37215
1559           Westminster         8105
1560           Westminster        25604
1561           Westminster           34
1562           Westminster          969
1563           Westminster         6503
1564           Westminster         1347
1565           Westminster            7
1566           Westminster         7478
1567           Westminster        13344

[1568 rows x 2 columns]

Grouping the crime data for each boroughs

In [11]:
df_crime = df_mps2.groupby(['Address'])['Total_Crime'].mean().reset_index()

In [12]:
df_crime.head()

Address  Total_Crime
0  Barking and Dagenham   809.312500
1                Barnet  1259.851064
2                Bexley   734.173913
3                 Brent  1244.416667
4               Bromley  1033.239130

Labelling each borough with London, UK in order to pull the correct coordinates

In [13]:
df_crime['Address'] = df_crime['Address'] + ', London, UK'
df_crime.head()

Address  Total_Crime
0  Barking and Dagenham, London, UK   809.312500
1                Barnet, London, UK  1259.851064
2                Bexley, London, UK   734.173913
3                 Brent, London, UK  1244.416667
4               Bromley, London, UK  1033.239130

In [14]:
import pandas as pd
import numpy as np
import datetime as DT
import hmac
from geopy.geocoders import Nominatim
# import k-means from clustering stage
from geopy.extra.rate_limiter import RateLimiter
from functools import partial
from sklearn.cluster import KMeans

Using the geopy library to get the coordinates of the London boroughs

In [15]:
geolocator = Nominatim(user_agent="ldn_explorer")
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)
# Create location column
df_crime['location'] = df_crime['Address'].apply(geocode)
df_crime['point']=df_crime['location'].apply(lambda loc: tuple(loc.point) if loc else None)
df_crime.head()

Address  Total_Crime  \
0  Barking and Dagenham, London, UK   809.312500   
1                Barnet, London, UK  1259.851064   
2                Bexley, London, UK   734.173913   
3                 Brent, London, UK  1244.416667   
4               Bromley, London, UK  1033.239130   

                                            location  \
0  (London Borough of Barking and Dagenham, Great...   
1  (Chipping Barnet, London Borough of Barnet, Lo...   
2  (Bexley, London Borough of Bexley, London, Gre...   
3  (London Borough of Brent, Greater London, Engl...   
4  (Bromley, London, Greater London, England, BR1...   

                                            point  
0          (51.5541171, 0.15050434261994267, 0.0)  
1                     (51.65309, -0.2002261, 0.0)  
2                     (51.4416793, 0.150488, 0.0)  
3  (51.563825800000004, -0.2757596561855699, 0.0)  
4                    (51.4028046, 0.0148142, 0.0)

Separating the coordinates into separate columns

In [16]:
df_crime[['Latitude', 'Longitude', 'Altitude']] = pd.DataFrame(df_crime['point'].tolist(), index=df_crime.index)
df_crime.head()

Address  Total_Crime  \
0  Barking and Dagenham, London, UK   809.312500   
1                Barnet, London, UK  1259.851064   
2                Bexley, London, UK   734.173913   
3                 Brent, London, UK  1244.416667   
4               Bromley, London, UK  1033.239130   

                                            location  \
0  (London Borough of Barking and Dagenham, Great...   
1  (Chipping Barnet, London Borough of Barnet, Lo...   
2  (Bexley, London Borough of Bexley, London, Gre...   
3  (London Borough of Brent, Greater London, Engl...   
4  (Bromley, London, Greater London, England, BR1...   

                                            point   Latitude  Longitude  \
0          (51.5541171, 0.15050434261994267, 0.0)  51.554117   0.150504   
1                     (51.65309, -0.2002261, 0.0)  51.653090  -0.200226   
2                     (51.4416793, 0.150488, 0.0)  51.441679   0.150488   
3  (51.563825800000004, -0.2757596561855699, 0.0)  51.563826  -0.275760   
4                    (51.4028046, 0.0148142, 0.0)  51.402805   0.014814   

   Altitude  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0

Dropping the unnecessary columns

In [17]:
df_crime2 = df_crime.drop(df_crime.columns[[2, 3, 6]], axis=1)
df_crime2.head()

Address  Total_Crime   Latitude  Longitude
0  Barking and Dagenham, London, UK   809.312500  51.554117   0.150504
1                Barnet, London, UK  1259.851064  51.653090  -0.200226
2                Bexley, London, UK   734.173913  51.441679   0.150488
3                 Brent, London, UK  1244.416667  51.563826  -0.275760
4               Bromley, London, UK  1033.239130  51.402805   0.014814

In [18]:
!pip install folium
import folium
print('Libraries imported.')

     |████████████████████████████████| 102kB 6.9MB/s ta 0:00:011
Libraries imported.


In [19]:
address = 'London, UK'

geolocator = Nominatim(user_agent="ldn_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of London City are 51.5073219, -0.1276474.


### Exploring neighborhoods in London

In [20]:
!pip install geopandas
!pip install geopy

     |████████████████████████████████| 972kB 6.7MB/s eta 0:00:01
     |████████████████████████████████| 14.7MB 20.0MB/s eta 0:00:01    |██████████████████▏             | 8.4MB 20.0MB/s eta 0:00:01
     |████████████████████████████████| 1.8MB 26.2MB/s eta 0:00:01
     |████████████████████████████████| 10.9MB 10.6MB/s eta 0:00:01     |████████████████████████████████| 10.9MB 10.6MB/s eta 0:00:01


Using the Foursquare API to explore the neighborhoods and segment them

In [21]:
#Define Foursquare Credentials and Version

CLIENT_ID = 'CLQJQRLVWRV1NPPLLMIDF5OZHFP1W3HZXXJSO1JAVI0UXCSG' # your Foursquare ID
CLIENT_SECRET = 'QNXR2VJSFYIW4KJFWQMGFXBBQAQZJTS40PVODGVTXDL45YKO' # your Foursquare Secret
VERSION = '20180605'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CLQJQRLVWRV1NPPLLMIDF5OZHFP1W3HZXXJSO1JAVI0UXCSG
CLIENT_SECRET:QNXR2VJSFYIW4KJFWQMGFXBBQAQZJTS40PVODGVTXDL45YKO


In [22]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [23]:
df_crime2.drop(df_crime2.index[22], inplace = True)

Creating a new dataframe for the London venues

In [24]:
london_venues = getNearbyVenues(names=df_crime2['Address'],
                                   latitudes=df_crime2['Latitude'],
                                   longitudes=df_crime2['Longitude']
                                  )

Barking and Dagenham, London, UK
Barnet, London, UK
Bexley, London, UK
Brent, London, UK
Bromley, London, UK
Camden, London, UK
Croydon, London, UK
Ealing, London, UK
Enfield, London, UK
Greenwich, London, UK
Hackney, London, UK
Hammersmith and Fulham, London, UK
Haringey, London, UK
Harrow, London, UK
Havering, London, UK
Hillingdon, London, UK
Hounslow, London, UK
Islington, London, UK
Kensington and Chelsea, London, UK
Kingston upon Thames, London, UK
Lambeth, London, UK
Lewisham, London, UK
Merton, London, UK
Newham, London, UK
Redbridge, London, UK
Richmond upon Thames, London, UK
Southwark, London, UK
Sutton, London, UK
Tower Hamlets, London, UK
Waltham Forest, London, UK
Wandsworth, London, UK
Westminster, London, UK


In [25]:
london_venues.head()

Neighborhood  Neighborhood Latitude  \
0  Barking and Dagenham, London, UK              51.554117   
1  Barking and Dagenham, London, UK              51.554117   
2  Barking and Dagenham, London, UK              51.554117   
3  Barking and Dagenham, London, UK              51.554117   
4  Barking and Dagenham, London, UK              51.554117   

   Neighborhood Longitude                     Venue  Venue Latitude  \
0                0.150504             Tesco Express       51.551536   
1                0.150504      Connor Road Bus Stop       51.554345   
2                0.150504  Oglethorpe Road Bus Stop       51.555221   
3                0.150504     Five Elms Off Licence       51.553878   
4                0.150504               Post office       51.551411   

   Venue Longitude     Venue Category  
0         0.152784      Grocery Store  
1         0.147162           Bus Stop  
2         0.147136           Bus Stop  
3         0.145531       Liquor Store  
4         0.155003  Convenience Store

Checking how many venues are returned for each neighborhood

In [26]:
london_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                
Barking and Dagenham, London, UK                        5   
Barnet, London, UK                                     31   
Bexley, London, UK                                     12   
Brent, London, UK                                      16   
Bromley, London, UK                                    44   
Camden, London, UK                                     86   
Croydon, London, UK                                    25   
Ealing, London, UK                                     95   
Enfield, London, UK                                    57   
Greenwich, London, UK                                  60   
Hackney, London, UK                                    31   
Hammersmith and Fulham, London, UK                     89   
Haringey, London, UK                                   32   
Harrow, London, UK                                      7   
Havering, London, UK                                   68   
Hillingdon, London, UK                                  4   
Hounslow, London, UK                                   51   
Islington, London, UK                                  73   
Kensington and Chelsea, London, UK                     43   
Kingston upon Thames, London, UK                      100   
Lambeth, London, UK                                    61   
Lewisham, London, UK                                   50   
Merton, London, UK                                     16   
Newham, London, UK                                      4   
Redbridge, London, UK                                   5   
Richmond upon Thames, London, UK                        5   
Southwark, London, UK                                 100   
Sutton, London, UK                                      5   
Tower Hamlets, London, UK                              43   
Waltham Forest, London, UK                              4   
Wandsworth, London, UK                                 37   
Westminster, London, UK                                45   

                                    Neighborhood Longitude  Venue  \
Neighborhood                                                        
Barking and Dagenham, London, UK                         5      5   
Barnet, London, UK                                      31     31   
Bexley, London, UK                                      12     12   
Brent, London, UK                                       16     16   
Bromley, London, UK                                     44     44   
Camden, London, UK                                      86     86   
Croydon, London, UK                                     25     25   
Ealing, London, UK                                      95     95   
Enfield, London, UK                                     57     57   
Greenwich, London, UK                                   60     60   
Hackney, London, UK                                     31     31   
Hammersmith and Fulham, London, UK                      89     89   
Haringey, London, UK                                    32     32   
Harrow, London, UK                                       7      7   
Havering, London, UK                                    68     68   
Hillingdon, London, UK                                   4      4   
Hounslow, London, UK                                    51     51   
Islington, London, UK                                   73     73   
Kensington and Chelsea, London, UK                      43     43   
Kingston upon Thames, London, UK                       100    100   
Lambeth, London, UK                                     61     61   
Lewisham, London, UK                                    50     50   
Merton, London, UK                                      16     16   
Newham, London, UK                                       4      4   
Redbridge, London, UK                                    5      5   
Richmond upon Thames, London, UK                         5      5   
Southwark, London, UK        

### Analysing each neighborhood

In [27]:
# one hot encoding
london_onehot = pd.get_dummies(london_venues[['Venue Category']], prefix="", prefix_sep="")

# add street column back to dataframe
london_onehot['Neighborhood'] = london_venues['Neighborhood'] 

# move borough column to the first column
fixed_columns = [london_onehot.columns[-1]] + list(london_onehot.columns[:-1])

#fixed_columns
london_onehot = london_onehot[fixed_columns]

london_onehot.head()

Neighborhood  Afghan Restaurant  African Restaurant  \
0  Barking and Dagenham, London, UK                  0                   0   
1  Barking and Dagenham, London, UK                  0                   0   
2  Barking and Dagenham, London, UK                  0                   0   
3  Barking and Dagenham, London, UK                  0                   0   
4  Barking and Dagenham, London, UK                  0                   0   

   Aquarium  Arcade  Argentinian Restaurant  Art Gallery  Arts & Crafts Store  \
0         0       0                       0            0                    0   
1         0       0                       0            0                    0   
2         0       0                       0            0                    0   
3         0       0                       0            0                    0   
4         0       0                       0            0                    0   

   Asian Restaurant  Athletics & Sports  ...  Turkish Restaurant  \
0                 0                   0  ...                   0   
1                 0                   0  ...                   0   
2                 0                   0  ...                   0   
3                 0                   0  ...                   0   
4                 0                   0  ...                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 227 columns]

Grouping the rows by neighborhood and the mean of frequency of occurrence of each category

In [28]:
london_grouped = london_onehot.groupby('Neighborhood').mean().reset_index()
london_grouped.head()

Neighborhood  Afghan Restaurant  African Restaurant  \
0  Barking and Dagenham, London, UK                0.0                 0.0   
1                Barnet, London, UK                0.0                 0.0   
2                Bexley, London, UK                0.0                 0.0   
3                 Brent, London, UK                0.0                 0.0   
4               Bromley, London, UK                0.0                 0.0   

   Aquarium  Arcade  Argentinian Restaurant  Art Gallery  Arts & Crafts Store  \
0       0.0     0.0                     0.0          0.0                  0.0   
1       0.0     0.0                     0.0          0.0                  0.0   
2       0.0     0.0                     0.0          0.0                  0.0   
3       0.0     0.0                     0.0          0.0                  0.0   
4       0.0     0.0                     0.0          0.0                  0.0   

   Asian Restaurant  Athletics & Sports  ...  Turkish Restaurant  \
0          0.000000                 0.0  ...            0.000000   
1          0.000000                 0.0  ...            0.000000   
2          0.000000                 0.0  ...            0.000000   
3          0.000000                 0.0  ...            0.000000   
4          0.022727                 0.0  ...            0.022727   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                            0.0               0.0                    0.0   
1                            0.0               0.0                    0.0   
2                            0.0               0.0                    0.0   
3                            0.0               0.0                    0.0   
4                            0.0               0.0                    0.0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0              0.0       0.0        0.0          0.0            0.0   
1              0.0       0.0        0.0          0.0            0.0   
2              0.0       0.0        0.0          0.0            0.0   
3              0.0       0.0        0.0          0.0            0.0   
4              0.0       0.0        0.0          0.0            0.0   

   Yoga Studio  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 227 columns]

Writing a function to sort the venues in descending order

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Creating a new dataframe and displaying the top 10 venues for each neighborhood

In [31]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = london_grouped['Neighborhood']

for ind in np.arange(london_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(london_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Barking and Dagenham, London, UK              Bus Stop   
1                Barnet, London, UK           Coffee Shop   
2                Bexley, London, UK                   Pub   
3                 Brent, London, UK           Coffee Shop   
4               Bromley, London, UK        Clothing Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Grocery Store     Convenience Store          Liquor Store   
1              Pharmacy             Bookstore                  Park   
2  Fast Food Restaurant      Greek Restaurant    Chinese Restaurant   
3                 Hotel           Supermarket     Indian Restaurant   
4           Coffee Shop  Gym / Fitness Center           Pizza Place   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0           Event Space      Food & Drink Shop                  Food   
1     Convenience Store            Pizza Place            Restaurant   
2      Toy / Game Store          Train Station          Tennis Court   
3      Pedestrian Plaza      Electronics Store        Sandwich Place   
4                   Pub  Portuguese Restaurant          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Flea Market           Fish Market      Fish & Chips Shop  
1         Grocery Store                   Pub      Mobile Phone Shop  
2    Italian Restaurant        Breakfast Spot      Indian Restaurant  
3                  Café          Burger Joint             Food Court  
4             Bookstore      Stationery Store            Supermarket

### Clustering neighborhoods

Running *k*-means to cluster the neighborhood into 5 clusters

In [32]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

london_grouped_clustering = london_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(london_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 2, 0, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

Creating a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood 

In [33]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

london_merged = df_crime2

# merge london_grouped with df_crime2 to add latitude/longitude for each neighborhood
london_merged = london_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Address')

london_merged.head()

Address  Total_Crime   Latitude  Longitude  \
0  Barking and Dagenham, London, UK   809.312500  51.554117   0.150504   
1                Barnet, London, UK  1259.851064  51.653090  -0.200226   
2                Bexley, London, UK   734.173913  51.441679   0.150488   
3                 Brent, London, UK  1244.416667  51.563826  -0.275760   
4               Bromley, London, UK  1033.239130  51.402805   0.014814   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1              Bus Stop         Grocery Store   
1               2           Coffee Shop              Pharmacy   
2               0                   Pub  Fast Food Restaurant   
3               2           Coffee Shop                 Hotel   
4               2        Clothing Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Convenience Store          Liquor Store           Event Space   
1             Bookstore                  Park     Convenience Store   
2      Greek Restaurant    Chinese Restaurant      Toy / Game Store   
3           Supermarket     Indian Restaurant      Pedestrian Plaza   
4  Gym / Fitness Center           Pizza Place                   Pub   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Food & Drink Shop                  Food           Flea Market   
1            Pizza Place            Restaurant         Grocery Store   
2          Train Station          Tennis Court    Italian Restaurant   
3      Electronics Store        Sandwich Place                  Café   
4  Portuguese Restaurant          Burger Joint             Bookstore   

  9th Most Common Venue 10th Most Common Venue  
0           Fish Market      Fish & Chips Shop  
1                   Pub      Mobile Phone Shop  
2        Breakfast Spot      Indian Restaurant  
3          Burger Joint             Food Court  
4      Stationery Store            Supermarket

In [34]:
neighborhoods_venues_sorted.head()

Cluster Labels                      Neighborhood 1st Most Common Venue  \
0               1  Barking and Dagenham, London, UK              Bus Stop   
1               2                Barnet, London, UK           Coffee Shop   
2               0                Bexley, London, UK                   Pub   
3               2                 Brent, London, UK           Coffee Shop   
4               2               Bromley, London, UK        Clothing Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Grocery Store     Convenience Store          Liquor Store   
1              Pharmacy             Bookstore                  Park   
2  Fast Food Restaurant      Greek Restaurant    Chinese Restaurant   
3                 Hotel           Supermarket     Indian Restaurant   
4           Coffee Shop  Gym / Fitness Center           Pizza Place   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0           Event Space      Food & Drink Shop                  Food   
1     Convenience Store            Pizza Place            Restaurant   
2      Toy / Game Store          Train Station          Tennis Court   
3      Pedestrian Plaza      Electronics Store        Sandwich Place   
4                   Pub  Portuguese Restaurant          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Flea Market           Fish Market      Fish & Chips Shop  
1         Grocery Store                   Pub      Mobile Phone Shop  
2    Italian Restaurant        Breakfast Spot      Indian Restaurant  
3                  Café          Burger Joint             Food Court  
4             Bookstore      Stationery Store            Supermarket

Visualising the resulting clusters

In [35]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_merged['Latitude'], london_merged['Longitude'], london_merged['Address'], london_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining Clusters

#### Cluster 1

In [36]:
london_merged.loc[london_merged['Cluster Labels'] == 0, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Total_Crime 1st Most Common Venue 2nd Most Common Venue  \
2    734.173913                   Pub  Fast Food Restaurant   
15  1092.583333                   Pub  Fast Food Restaurant   
23   627.311111          Tram Station                   Pub   
24  1490.021277                   Pub           Bus Station   
30  1004.812500                   Pub              Bus Stop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2       Greek Restaurant    Chinese Restaurant      Toy / Game Store   
15                  Park    Chinese Restaurant           Yoga Studio   
23                  Park          Dessert Shop                  Farm   
24                  Café                  Park           Yoga Studio   
30                  Café           Auto Garage           Yoga Studio   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2          Train Station          Tennis Court    Italian Restaurant   
15    English Restaurant     Food & Drink Shop                  Food   
23      Sushi Restaurant  Fast Food Restaurant            Steakhouse   
24    Falafel Restaurant     Food & Drink Shop                  Food   
30    Falafel Restaurant     Food & Drink Shop                  Food   

   9th Most Common Venue 10th Most Common Venue  
2         Breakfast Spot      Indian Restaurant  
15           Flea Market            Fish Market  
23       Thai Restaurant         Hardware Store  
24           Flea Market            Fish Market  
30           Flea Market            Fish Market

#### Cluster 2

In [37]:
london_merged.loc[london_merged['Cluster Labels'] == 1, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Total_Crime 1st Most Common Venue 2nd Most Common Venue  \
0     809.3125              Bus Stop         Grocery Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Convenience Store          Liquor Store           Event Space   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Food & Drink Shop                  Food           Flea Market   

  9th Most Common Venue 10th Most Common Venue  
0           Fish Market      Fish & Chips Shop

#### Cluster 3

In [38]:
london_merged.loc[london_merged['Cluster Labels'] == 2, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Total_Crime 1st Most Common Venue 2nd Most Common Venue  \
1   1259.851064           Coffee Shop              Pharmacy   
3   1244.416667           Coffee Shop                 Hotel   
4   1033.239130        Clothing Store           Coffee Shop   
5   1441.280000                   Pub           Coffee Shop   
6   1356.448980                   Pub           Coffee Shop   
7   1262.000000           Coffee Shop                   Pub   
8   1223.708333        Clothing Store           Coffee Shop   
9   1120.530612                   Pub         Boat or Ferry   
10  1299.200000           Coffee Shop                   Pub   
11   921.416667                   Pub                  Café   
12  1256.040816              Bus Stop                  Park   
13   670.102041     Afghan Restaurant           Coffee Shop   
14   778.212766                  Café                   Pub   
16  1037.580000           Coffee Shop                 Hotel   
17  1185.000000                   Pub          Burger Joint   
18   939.520833                  Café          Cupcake Shop   
19   561.644444           Coffee Shop                  Café   
20  1392.102041           Coffee Shop                 Hotel   
21  1143.734694        Clothing Store  Fast Food Restaurant   
25   962.469388                 Hotel                   Pub   
27  1467.000000                 Hotel                   Pub   
29  1369.285714                   Pub           Pizza Place   
31  1051.812500        Clothing Store           Coffee Shop   
32  2868.591837           Coffee Shop         Historic Site   

   3rd Most Common Venue        4th Most Common Venue  \
1              Bookstore                         Park   
3            Supermarket            Indian Restaurant   
4   Gym / Fitness Center                  Pizza Place   
5           Burger Joint           Italian Restaurant   
6     Spanish Restaurant           Italian Restaurant   
7         Clothing Store                 Burger Joint   
8                    Pub                  Supermarket   
9           Burger Joint               History Museum   
10           Supermarket                      Brewery   
11           Coffee Shop         Gym / Fitness Center   
12                  Café         Fast Food Restaurant   
13  Fast Food Restaurant               Sandwich Place   
14           Coffee Shop                          Bar   
16         Grocery Store               Clothing Store   
17                Bakery            French Restaurant   
18          Burger Joint           Italian Restaurant   
19                   Pub           Italian Restaurant   
20                   Bar                         Café   
21           Coffee Shop                Grocery Store   
25         Metro Station  Eastern European Restaurant   
27  Gym / Fitness Center                  Coffee Shop   
29           Coffee Shop                     Bus Stop   
31                   Pub                 Burger Joint   
32     Outdoor Sculpture               Sandwich Place   

       5th Most Common Venue     6th Most Common Venue  \
1          Convenience Store               Pizza Place   
3           Pedestrian Plaza         Electronics Store   
4                        Pub     Portuguese Restaurant   
5                       Café                    Market   
6               Burger Joint  Mediterranean Restaurant   
7                     Bakery                  Platform   
8                       Café                 Gift Shop   
9                     Garden                    Market   
10                      Café              Cocktail Bar   
11            Sandwich Place                     Hotel   
12               Coffee Shop          Malay Restaurant   
13            Clothing Store             Grocery Store   
14                    Market               Flea Market   
16      Fast Food Restaurant         Indian Restaurant   
17  Mediterranean Restaurant              Cocktail Bar   
18               Supermarket                    Bakery   
19            Cl

#### Cluster 4

In [39]:
london_merged.loc[london_merged['Cluster Labels'] == 3, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Total_Crime 1st Most Common Venue 2nd Most Common Venue  \
28   610.318182                Bakery         Train Station   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
28                  Park    Italian Restaurant         Grocery Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
28           Flea Market           Fish Market     Fish & Chips Shop   

   9th Most Common Venue 10th Most Common Venue  
28   Filipino Restaurant     English Restaurant

#### Cluster 5

In [40]:
london_merged.loc[london_merged['Cluster Labels'] == 4, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Total_Crime 1st Most Common Venue 2nd Most Common Venue  \
26   590.116279           Sports Club          Home Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
26                   Pub    Photography Studio           Bus Station   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
26    English Restaurant                  Food           Flea Market   

   9th Most Common Venue 10th Most Common Venue  
26           Fish Market      Fish & Chips Shop

### Summary of crime and venue data

In [41]:
london_merged.sort_values(by=['Total_Crime','Cluster Labels'], inplace=True)
london_merged

Address  Total_Crime   Latitude  Longitude  \
19    Kingston upon Thames, London, UK   561.644444  51.409627  -0.306262   
26    Richmond upon Thames, London, UK   590.116279  51.440372  -0.305720   
28                  Sutton, London, UK   610.318182  51.357511  -0.173640   
23                  Merton, London, UK   627.311111  51.410803  -0.188099   
13                  Harrow, London, UK   670.102041  51.596827  -0.337316   
2                   Bexley, London, UK   734.173913  51.441679   0.150488   
14                Havering, London, UK   778.212766  51.544369  -0.144303   
0     Barking and Dagenham, London, UK   809.312500  51.554117   0.150504   
11  Hammersmith and Fulham, London, UK   921.416667  51.492038  -0.223640   
18  Kensington and Chelsea, London, UK   939.520833  51.498995  -0.199123   
25               Redbridge, London, UK   962.469388  51.576320   0.045410   
30          Waltham Forest, London, UK  1004.812500  51.598169  -0.017837   
4                  Bromley, London, UK  1033.239130  51.402805   0.014814   
16                Hounslow, London, UK  1037.580000  51.468613  -0.361347   
31              Wandsworth, London, UK  1051.812500  51.457027  -0.193261   
15              Hillingdon, London, UK  1092.583333  51.542519  -0.448335   
9                Greenwich, London, UK  1120.530612  51.482084  -0.004542   
21                Lewisham, London, UK  1143.734694  51.462432  -0.010133   
17               Islington, London, UK  1185.000000  51.538429  -0.099905   
8                  Enfield, London, UK  1223.708333  51.652085  -0.081018   
3                    Brent, London, UK  1244.416667  51.563826  -0.275760   
12                Haringey, London, UK  1256.040816  51.587930  -0.105410   
1                   Barnet, London, UK  1259.851064  51.653090  -0.200226   
7                   Ealing, London, UK  1262.000000  51.512655  -0.305195   
10                 Hackney, London, UK  1299.200000  51.543240  -0.049362   
6                  Croydon, London, UK  1356.448980  51.371305  -0.101957   
29           Tower Hamlets, London, UK  1369.285714  51.525629  -0.033585   
20                 Lambeth, London, UK  1392.102041  51.501301  -0.117287   
5                   Camden, London, UK  1441.280000  51.542305  -0.139560   
27               Southwark, London, UK  1467.000000  51.502922  -0.103458   
24                  Newham, London, UK  1490.021277  51.530000   0.029318   
32             Westminster, London, UK  2868.591837  51.500444  -0.126540   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19               2           Coffee Shop                  Café   
26               4           Sports Club          Home Service   
28               3                Bakery         Train Station   
23               0          Tram Station                   Pub   
13               2     Afghan Restaurant           Coffee Shop   
2                0                   Pub  Fast Food Restaurant   
14               2                  Café                   Pub   
0                1              Bus Stop         Grocery Store   
11               2                   Pub                  Café   
18               2                  Café          Cupcake Shop   
25               2                 Hotel                   Pub   
30               0                   Pub              Bus Stop   
4                2        Clothing Store           Coffee Shop   
16               2           Coffee Shop                 Hotel   
31               2        Clothing Store           Coffee Shop   
15               0                   Pub  Fast Food Restaurant   
9                2                   Pub         Boat or Ferry   
21               2        Clothing Store  Fast Food Restaurant   
17               2                   Pub          Burger Joint   
8                2        Clothing Store           Coffee Shop   
3                2           Coffee Shop                 Hotel   
12               2              Bus S

## Results and discussion

We analysed two sources of data in this report. Hence, we may discuss our results in two main parts. 

In terms of the Metropolitan Police Service Borough Level Crime data, we found that the boroughs with the least crime are: (1) Kingston upon Thames, (2) Richmond upon Thames and (3) Sutton. The boroughs with the most crime are: (1) Westminster, (2) Newham, and (3) Southwark. The quartiles for the crime data are: 893.4, 1106.6, 1271.3, 2868.6 crimes over the course of 24 months. The mean number of crimes is 1118.9. 

Clusters 2, 4 and 5 all showed crime levels below average. However, clusters 1 and 3 showed mixed crime levels across each borough. 

The most common venues were as follows:

Cluster 1 – Pubs, fast food restaurants, parks.

Cluster 2 – Bus stops, grocery stores, convenience stores. 

Cluster 3 – Coffee shops, pubs and clothing stores. 

Cluster 4 – Bakeries, train stations, parks. 

Cluster 5 – Sports clubs, home services, pubs. 

The clusters that would appeal most to families and older people, with fewer pubs and more active spaces and independent stores, are Cluster 4 and Cluster 5 as they offer suitable facilities and low crime rates. Hence, the boroughs of Sutton (4) and Richmond (5) are attractive for family homes and older people. 

Cluster 1 seems to be appropriate for younger residential buyers/renters with more pubs, fast food restaurants and parks. Within Cluster 1, the borough with the least crime is Merton, followed by Bexley. 

Cluster 3 would be fitting for either group, especially families with slightly older children, with a variety of coffee shops, pubs and clothing stores. Within this cluster, the boroughs with the least crime are Kingston upon Thames and Harrow. 

Finally, Cluster 4 appears to have low crime rates, but has much fewer desirable venues such as pubs or cafes. Hence, Barking and Dagenham may be safe but less attractive for buyers/renters. 

## Conclusion

The business question that we have answered in this report is: which parts of London are safe and attractive for residential buyers/renters across various demographic groups? To solve this problem, we have assessed the crime rates in each of the London boroughs, and clustered neighborhoods to assess the venues on offer, such as pubs, cafes and parks.

Overall, the ideal borough for a residential buyer/renter depends on the age and venue preference of the individual. However, it seems that the Sutton and Richmond are most suitable for older people or families with young children. Kingston upon Thames and Harrow are ideal for families with older childer. Merton and Bexley are appropriate for young people looking for more pubs and fast food restaurant. Barking and Dagenham is a safe option but with few desirable venues.

Whilst London crime rates have been growing over the last few years, there are many boroughs which are both relatively safe and offer venues which are appealing to all ages. This report acts as a guide to all commercial and residential real estate agents, as well as anybody looking to buy or rent a home in the city, and should inform individuals across all ages. 